refactor(ast): `BoundNames` trait should not allocate

The `BoundNames` trait currently allocates

https://github.com/Boshen/oxc/blob/249cf96ad270c6a81bea228ae3b568dbfe3898e4/crates/oxc_ast/src/syntax_directed_operations.rs#L12

We should change this to a procedure which not does allocate

I tried to pass in a closure for the trait and failed.

[bound_names](https://github.com/Boshen/oxc/tree/bound_names)

```
pub trait BoundNames {
    fn bound_names<F>(&self, f: F)
    where
        F: Fn(&BindingIdentifier) -> ();
}
```

This needs to be a `FnMut`, 

```
error[E0596]: cannot borrow `*builder` as mutable, as it is a captured variable in a `Fn` closure
   --> crates/oxc_semantic/src/binder.rs:137:17
    |
137 | /                 builder.declare_symbol(
138 | |                     &ident.name,
139 | |                     ident.span,
140 | |                     builder.scope.current_scope_id,
141 | |                     includes,
142 | |                     excludes,
143 | |                 );
    | |_________________^ cannot borrow as mutable
```

but I can't get it to compile 😢

--- a/crates/oxc_ast/src/syntax_directed_operations.rs
+++ b/crates/oxc_ast/src/syntax_directed_operations.rs
@@ -5,158 +5,150 @@ use crate::{ast::*, Span};
 
 /// [`BoundName`](https://tc39.es/ecma262/#sec-static-semantics-boundnames)
 pub trait BoundName {
-    fn bound_name(&self) -> Option<&BindingIdentifier>;
+    fn bound_name<F: FnMut(&BindingIdentifier)>(&self, f: &mut F);
 }
 
 pub trait BoundNames {
-    fn bound_names(&self) -> Vec<&BindingIdentifier>;
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F);
 }
 
 impl<'a> BoundNames for BindingPattern<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
         match &self.kind {
-            BindingPatternKind::BindingIdentifier(ident) => ident.bound_names(),
-            BindingPatternKind::ArrayPattern(array) => array.bound_names(),
-            BindingPatternKind::ObjectPattern(object) => object.bound_names(),
-            BindingPatternKind::AssignmentPattern(assignment) => assignment.bound_names(),
-            BindingPatternKind::RestElement(rest) => rest.argument.bound_names(),
+            BindingPatternKind::BindingIdentifier(ident) => ident.bound_names(f),
+            BindingPatternKind::ArrayPattern(array) => array.bound_names(f),
+            BindingPatternKind::ObjectPattern(object) => object.bound_names(f),
+            BindingPatternKind::AssignmentPattern(assignment) => assignment.bound_names(f),
+            BindingPatternKind::RestElement(rest) => rest.argument.bound_names(f),
         }
     }
 }
 
-impl<'a> BoundNames for Option<BindingPattern<'a>> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.as_ref().map_or(vec![], BoundNames::bound_names)
-    }
-}
-
-impl BoundNames for Option<BindingIdentifier> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.as_ref().map_or(vec![], |ident| vec![ident])
-    }
-}
-
 impl BoundNames for BindingIdentifier {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        vec![self]
+    fn bound_names<F: FnMut(&Self)>(&self, f: &mut F) {
+        f(self);
     }
 }
 
 impl<'a> BoundNames for ArrayPattern<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.elements.iter().flat_map(BoundNames::bound_names).collect()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        for elem in self.elements.iter().flatten() {
+            elem.bound_names(f);
+        }
     }
 }
 
 impl<'a> BoundNames for ObjectPattern<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.properties
-            .iter()
-            .flat_map(|p| match p {
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        for p in &self.properties {
+            match p {
                 ObjectPatternProperty::Property(property) => {
                     if let PropertyValue::Pattern(pattern) = &property.value {
-                        pattern.bound_names()
-                    } else {
-                        vec![]
+                        pattern.bound_names(f);
                     }
                 }
-                ObjectPatternProperty::RestElement(rest) => rest.argument.bound_names(),
-            })
-            .collect()
+                ObjectPatternProperty::RestElement(rest) => rest.argument.bound_names(f),
+            }
+        }
     }
 }
 
 impl<'a> BoundNames for AssignmentPattern<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.left.bound_names()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.left.bound_names(f);
     }
 }
 
 impl<'a> BoundNames for RestElement<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.argument.bound_names()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.argument.bound_names(f);
     }
 }
 
 impl<'a> BoundNames for FormalParameters<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.items.iter().flat_map(BoundNames::bound_names).collect()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        for item in &self.items {
+            item.bound_names(f);
+        }
     }
 }
 
 impl<'a> BoundNames for Declaration<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
         match self {
-            Declaration::VariableDeclaration(decl) => decl.bound_names(),
-            Declaration::FunctionDeclaration(func) => func.bound_names(),
-            Declaration::ClassDeclaration(decl) => decl.bound_names(),
-            _ => vec![],
+            Declaration::VariableDeclaration(decl) => decl.bound_names(f),
+            Declaration::FunctionDeclaration(func) => func.bound_names(f),
+            Declaration::ClassDeclaration(decl) => decl.bound_names(f),
+            _ => {}
         }
     }
 }
 
 impl<'a> BoundNames for VariableDeclaration<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.declarations.iter().flat_map(|declarator| declarator.id.bound_names()).collect()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        for declarator in &self.declarations {
+            declarator.id.bound_names(f);
+        }
     }
 }
 
 impl<'a> BoundName for Function<'a> {
-    fn bound_name(&self) -> Option<&BindingIdentifier> {
-        self.id.as_ref()
+    fn bound_name<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        if let Some(ident) = &self.id {
+            f(ident);
+        }
     }
 }
 
 impl<'a> BoundNames for Function<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.bound_name().map_or(vec![], |name| vec![name])
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.bound_name(f);
     }
 }
 
 impl<'a> BoundName for Class<'a> {
-    fn bound_name(&self) -> Option<&BindingIdentifier> {
-        self.id.as_ref()
+    fn bound_name<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        if let Some(ident) = &self.id {
+            f(ident);
+        }
     }
 }
 
 impl<'a> BoundNames for Class<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.bound_name().map_or(vec![], |name| vec![name])
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.bound_name(f);
     }
 }
 
 impl<'a> BoundNames for FormalParameter<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.pattern.bound_names()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.pattern.bound_names(f);
     }
 }
 
 impl<'a> BoundNames for ModuleDeclaration<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        match &self.kind {
-            ModuleDeclarationKind::ImportDeclaration(decl) => decl.bound_names(),
-            // ModuleDeclarationKind::ExportNamedDeclaration(decl) => decl.bound_names(),
-            _ => vec![],
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        if let ModuleDeclarationKind::ImportDeclaration(decl) = &self.kind {
+            decl.bound_names(f);
         }
     }
 }
 
 impl<'a> BoundNames for ImportDeclaration<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.specifiers
-            .iter()
-            .map(|specifier| match specifier {
-                ImportDeclarationSpecifier::ImportSpecifier(specifier) => &specifier.local,
-                ImportDeclarationSpecifier::ImportDefaultSpecifier(specifier) => &specifier.local,
-                ImportDeclarationSpecifier::ImportNamespaceSpecifier(specifier) => &specifier.local,
-            })
-            .collect()
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        self.specifiers.iter().for_each(|specifier| match specifier {
+            ImportDeclarationSpecifier::ImportSpecifier(specifier) => f(&specifier.local),
+            ImportDeclarationSpecifier::ImportDefaultSpecifier(specifier) => f(&specifier.local),
+            ImportDeclarationSpecifier::ImportNamespaceSpecifier(specifier) => f(&specifier.local),
+        });
     }
 }
 
 impl<'a> BoundNames for ExportNamedDeclaration<'a> {
-    fn bound_names(&self) -> Vec<&BindingIdentifier> {
-        self.declaration.as_ref().map_or(vec![], BoundNames::bound_names)
+    fn bound_names<F: FnMut(&BindingIdentifier)>(&self, f: &mut F) {
+        if let Some(decl) = &self.declaration {
+            decl.bound_names(f);
+        }
     }
 }
 

--- a/crates/oxc_linter/src/rules/early_error/javascript.rs
+++ b/crates/oxc_linter/src/rules/early_error/javascript.rs
@@ -11,6 +11,7 @@ use oxc_diagnostics::{
     Redeclaration,
 };
 use oxc_semantic::ScopeFlags;
+use rustc_hash::FxHashMap;
 
 use crate::{ast_util::STRICT_MODE_NAMES, context::LintContext, rule::Rule, AstNode};
 
@@ -133,15 +134,12 @@ fn check_program(node: &AstNode, ctx: &LintContext) {
 }
 
 fn check_duplicate_bound_names<T: BoundNames>(bound_names: &T, ctx: &LintContext) {
-    // bound_names are usually small, a simple loop should be more performant checking with a hashmap
-    let mut idents = bound_names.bound_names();
-    idents.sort_unstable_by_key(|ident| ident.name.as_str());
-    for i in 1..idents.len() {
-        let ident = &idents[i - 1];
-        if let Some(found) = idents[i..].iter().find(|i| i.name == ident.name) {
-            ctx.diagnostic(Redeclaration(ident.name.clone(), ident.span, found.span));
+    let mut idents: FxHashMap<Atom, Span> = FxHashMap::default();
+    bound_names.bound_names(&mut |ident| {
+        if let Some(old_span) = idents.insert(ident.name.clone(), ident.span) {
+            ctx.diagnostic(Redeclaration(ident.name.clone(), old_span, ident.span));
         }
-    }
+    });
 }
 
 #[derive(Debug, Error, Diagnostic)]

--- a/crates/oxc_semantic/src/binder.rs
+++ b/crates/oxc_semantic/src/binder.rs
@@ -29,7 +29,7 @@ impl<'a> Binder for VariableDeclarator<'a> {
                 (SymbolFlags::FunctionScopedVariable, SymbolFlags::FunctionScopedVariableExcludes)
             }
         };
-        for ident in self.id.bound_names() {
+        self.id.bound_names(&mut |ident| {
             let symbol_id = builder.declare_symbol(
                 &ident.name,
                 ident.span,
@@ -60,7 +60,7 @@ impl<'a> Binder for VariableDeclarator<'a> {
                     }
                 }
             }
-        }
+        });
     }
 }
 
@@ -132,7 +132,7 @@ impl<'a> Binder for FormalParameters<'a> {
         let includes = SymbolFlags::FunctionScopedVariable;
         let excludes = SymbolFlags::FunctionScopedVariableExcludes;
         let is_signature = self.kind == FormalParameterKind::Signature;
-        for ident in self.bound_names() {
+        self.bound_names(&mut |ident| {
             if !is_signature {
                 builder.declare_symbol(
                     &ident.name,
@@ -142,7 +142,7 @@ impl<'a> Binder for FormalParameters<'a> {
                     excludes,
                 );
             }
-        }
+        });
     }
 }
 
@@ -157,7 +157,7 @@ impl<'a> Binder for CatchClause<'a> {
                 let includes = SymbolFlags::FunctionScopedVariable | SymbolFlags::CatchVariable;
                 builder.declare_shadow_symbol(&ident.name, ident.span, current_scope_id, includes);
             } else {
-                for ident in param.bound_names() {
+                param.bound_names(&mut |ident| {
                     builder.declare_symbol(
                         &ident.name,
                         ident.span,
@@ -165,7 +165,7 @@ impl<'a> Binder for CatchClause<'a> {
                         SymbolFlags::BlockScopedVariable | SymbolFlags::CatchVariable,
                         SymbolFlags::BlockScopedVariableExcludes,
                     );
-                }
+                });
             }
         }
     }
@@ -173,7 +173,7 @@ impl<'a> Binder for CatchClause<'a> {
 
 impl<'a> Binder for ModuleDeclaration<'a> {
     fn bind(&self, builder: &mut SemanticBuilder) {
-        for ident in self.bound_names() {
+        self.bound_names(&mut |ident| {
             builder.declare_symbol(
                 &ident.name,
                 ident.span,
@@ -181,6 +181,6 @@ impl<'a> Binder for ModuleDeclaration<'a> {
                 SymbolFlags::empty(),
                 SymbolFlags::empty(),
             );
-        }
+        });
     }
 }

--- a/crates/oxc_semantic/src/module_record/builder.rs
+++ b/crates/oxc_semantic/src/module_record/builder.rs
@@ -256,7 +256,7 @@ impl ModuleRecordBuilder {
         }
 
         if let Some(decl) = &decl.declaration {
-            for ident in decl.bound_names() {
+            decl.bound_names(&mut |ident| {
                 let export_entry = ExportEntry {
                     module_request: module_request.clone(),
                     export_name: ExportExportName::Name(NameSpan::new(
@@ -271,7 +271,7 @@ impl ModuleRecordBuilder {
                 };
                 self.add_export_entry(export_entry);
                 self.add_export_binding(ident.name.clone(), ident.span);
-            }
+            });
         }
 
         for specifier in &decl.specifiers {

